### DISCUSSION TITLE
Bahas Fitur keamanan GPG di Github

### DISCUSSION THREAD

**Discussion by Andi-IM** [Andi-IM/desktop-project#13] [Show and tell]
Jadi gw ada bahas tentang keamanan menggunakan GPG.
> Sebelumnya kita juga pernah membahas menggunakan SSH (_Secure Shell_), silakan cek kembali issue #3. SSH juga mengamankan koneksi dari git menuju github. 

![alt text](https://user-images.githubusercontent.com/21165698/110954660-5417e400-837b-11eb-8e6f-0c392f5d3401.png "Keys")

Apa kegunaannya?
kita menggunakan GPG (_GNU Privacy Guard_) untuk membuat koneksi dari git di komputer menuju github menjadi lebih aman. Karena isi dari commit itu telah di enkripsi dengan berbagai macam format.
![image](https://user-images.githubusercontent.com/21165698/110955366-1a93a880-837c-11eb-8899-e8cc4a5bd738.png)

Pada catatan log akan terlihat verified seperti berikut.  
![image](https://user-images.githubusercontent.com/21165698/110954932-a2c57e00-837b-11eb-99cd-44eebab36e1b.png)

Langkah untuk konfigurasinya adalah sebagai berikut:
1. Download dan install [GPG command line tools ](https://gpg4win.org/download.html) ke sistem operasi yang digunakan. Kita secara umum merekomendasikan menginstall versi terbaru untuk sistem operasi Anda.
2. Buka git bash (tinggal cari aja di fitur _search_).
3. ketik perintah ini untuk generate GPG Key
```
$ gpg --full-generate-key
```
![image](https://user-images.githubusercontent.com/21165698/110956706-7dd20a80-837d-11eb-9a3a-3a8513ad05d5.png)

5. Pada prompt, pilih jenis key yang diinginkan, atau tekan `Enter` untuk menyetujui default `RSA and RSA`.
6. Masukkan besaran kunci, seharusnya lebih dari atau sama dengan `4096`.
![image](https://user-images.githubusercontent.com/21165698/110956842-a0642380-837d-11eb-8830-735f6f429e54.png)

7. Masukkan durasi waktu kunci berlaku. Tekan `Enter` untuk pengaturan default (tidak ada masa berlaku) **Tidak disarankan, lebih baik tentukan masa berlaku kuncinya**.
![image](https://user-images.githubusercontent.com/21165698/110956893-af4ad600-837d-11eb-97fa-f34acfe9999e.png)

8. Masukkan informasi user ID.
![image](https://user-images.githubusercontent.com/21165698/110957162-f0db8100-837d-11eb-8931-fd268bdd61ce.png)

9. Akan muncul permintaan password kode, masukkan password untuk kunci ini (harap untuk ingat baik-baik).
10. Kode telah di generate
![image](https://user-images.githubusercontent.com/21165698/110957632-5fb8da00-837e-11eb-8a4a-9e71b714b35c.png)

12. Gunakan perintah `gpg --list-secret-keys --keyid-format LONG` untuk melihat daftar kunci GPG yang Anda punya baik itu kunci publik atau kunci privat. Kunci privat dibutuhkan untuk menandai commit atau tag.
```$ gpg --list-secret-keys --keyid-format LONG```

14. dari list kunci GPG, copy ID kunci GPG yang akan digunakan. Misalnya, ID kunci GPG-nya  adalah `3AA5C34371567BD2`
```
$ gpg --list-secret-keys --keyid-format LONG
/Users/hubot/.gnupg/secring.gpg
------------------------------------
sec   4096R/3AA5C34371567BD2 2016-03-10 [expires: 2017-03-10]
uid                          Hubot 
ssb   4096R/42B317FD4BA89E7A 2016-03-10
```

15. Paste teks dibawah, ganti dengan ID kunci GPG yang Anda gunakan, misalnya, ID kunci GPG-nya adalah `3AA5C34371567BD2`:
```
$ gpg --armor --export 3AA5C34371567BD2
# Prints the GPG key ID, in ASCII armor format
```
17. Copy kunci GPG Anda, dimulai dengan `-----BEGIN PGP PUBLIC KEY BLOCK-----` dan diakhiri dengan `-----END PGP PUBLIC KEY BLOCK-----`.
18. Buka [Setting SSH dan GPG](https://github.com/settings/keys) akun anda
19. Klik [**New GPG Key**](https://github.com/settings/gpg/new).
20. Pastekan keynya dan klik **Add GPG key**
![image](https://user-images.githubusercontent.com/21165698/110959505-57fa3500-8380-11eb-865a-e80f1ed3cc07.png)

22.  Anda telah membuat keamanan berganda untuk koneksi Anda dengan repository github!

Untuk informasi lainnya bisa [baca disini](https://docs.github.com/en/github/authenticating-to-github/generating-a-new-gpg-key)

**Comment by Andi-IM** [comment]
Berdasarkan artikel dari [blog.bsk.im](https://blog.bsk.im/2018/01/18/signing-git-commits-with-gpg-on-windows-feat-sourcetree/). Kita ada perlu beberapa tambahan konfigurasi untuk membuat kode kita berjalan.

Kembali kita buka Git Bash
Ketik perintah `gpg --list-secret-keys` untuk mendapatkan kunci GPG privat dan publik
```
$ gpg --list-secret-keys
C:\Users\bskim\.gnupg/secring.gpg  
---------------------------------
sec   4096R/C8528BF2 2018-01-18  
uid                  Bumsoo Kim <bskim45@gmail.com>  
ssb   4096R/8911C702 2018-01-18  
```

Dari daftar kunci GPG, salin ID kunci GPG yang akan digunakan. Misalnya pada contoh diatas, ID kunci (KEY_ID) GPG-nya adalah `C8528BF2`.

Now export your private key with command . Replace KEY_ID with your key id copied above. Your private key is saved to private.asc file.
Sekarang export kunci privat anda dengan perintah `gpg --export-secret-key -a {KEY_ID} > private.asc` ( segera pindahkan private.asc ke tempat aman agar tidak direkam oleh git).

Jalankan Kleopatra, klik 'Import' atau persingkat dengan `ctrl + I` dan cari file `private.asc` yang telah dipindahkan tersebut.
![image](https://user-images.githubusercontent.com/21165698/111019542-7731aa80-83f2-11eb-9d7a-34a8d4f3cfef.png)

Anda akan diminta untuk mengkonfirmasi bahwa file kunci tersebut adalah milik Anda. Tekan 'Yes':
![image](https://user-images.githubusercontent.com/21165698/111019559-929cb580-83f2-11eb-9c66-ddabba329f4f.png)

Kunci GPG-nya telah berhasil diimport:
![image](https://user-images.githubusercontent.com/21165698/111019569-a0ead180-83f2-11eb-89b8-98e4317a6065.png)

Sekarang kita konfigurasi git untuk menggukanan gpg4win
Sejauh ini, kita membuat/mengimpor kunci ke Kleopatra, yang merupakan key manager utama dari gpg4win. Kunci sudah siap, namun kita butuh untuk memberitahu Git untuk menggunakan gpg4win untuk menandai commit daripada menggunakan `gpg.exe` bawaan.

Buka kembali git bash, ketik perintah ini dengan mengganti KEY_ID dengan ID kunci GPG yang Anda miliki.
`$ git config --global user.signingkey KEY_ID`

FYI: KEY_ID juga dapat dilihat pada Jendela Kleopatra
![image](https://user-images.githubusercontent.com/21165698/111019881-992c2c80-83f4-11eb-8595-c0c5030384af.png)

Pada Git versi 2.0.0 atau setelahnya, Anda tidak perlu mengetik -S setiap commit lagi. Untuk mengkonfigurasikan Git client Anda untuk menandai commit dapat dilakukan secara otomatis by default pada local repository apapun dengan mengetik perintah :
```
$ git config --global commit.gpgsign true
You can omit --global flag if you want to set default only for certain local repository.
```

Akhirnya, kita membuat git menggunakan gpg.exe versi gpg4win 
```
$ git config --global gpg.program "C:/Program Files (x86)/GNU/GnuPG/gpg2.exe"
```
**Jika Anda menggunakan Windows versi 32-bit atau jika Anda memiliki lokasi yang lain untuk menginstall gpg4win , ganti dengan lokasi yang sesuai.**

Untuk membuktikannya bekerja dengan baik, coba buat suatu commit, seharunya Anda mendapatkan konfirmasi password:
![image](https://user-images.githubusercontent.com/21165698/111019985-430bb900-83f5-11eb-88b5-c82e310ea9d2.png)

Anda bisa memferivikasi apa yang telah commit setelah ditandai.
```
$ git log --show-signature -1
commit 92e5d9505a8e593beba40b6e3d82a2fff85bbd6a (HEAD -> master)  
gpg: Signature made 01/18/18 15:56:27 Korea Standard Time^M  
gpg:                using RSA key BE86FF03DCD709AE0C3B5EFEB64A2E17C8528BF2^M  
gpg:                issuer "bskim45@gmail.com"^M  
gpg: Good signature from "Bumsoo Kim <bskim45@gmail.com>" [ultimate]^M  
```

Dan juga,  cek apakah berjalan juga pada SourceTree:
![image](https://user-images.githubusercontent.com/21165698/111020001-6df60d00-83f5-11eb-9525-404e0e76d832.png)

Akhirnya commit berhasil
![image](https://user-images.githubusercontent.com/21165698/111020036-aa296d80-83f5-11eb-93a3-f3e133c8d08c.png)

**Comment by k1rana** [comment]
Ijin koreksi mas, kebetulan nemu pas search "github gpg"
Pernyataan "Karena isi dari commit itu telah di enkripsi dengan berbagai macam format." kurang tepat.
Tepatnya commit bukan dienkripsi, melainkan ditanda tangani. Fungsi signature tersebut diperuntukkan kalau commit ini memang benar dari orang tersebut, karena pada dasarnya kita bisa mengubah nama dan email per commit sesuka hati kita. 
Jadi pernyataan yang benar, "Karena commit tersebut telah ditandatangani dengan GPG untuk memverifikasi keasliannya."

**Reply by Andi-IM** [reply]
Wah. terima kasih sudah reach-out proyek kecil-kecilan saya 😄 
Nanti saya akan update thread ini sesuai dengan yang seharusnya.
